### DISCUSSION TITLE
is there a way to add torrent using GET url parameters?

### DISCUSSION THREAD

**Discussion by cyberhck** [transmission/transmission#5248] [Q&A]
So I've added a protocol handler my chrome, it'd be awesome if I could popup the add dialog with pre-filled magnet url value without having to type it in, it could look something like: <transmission-ip>:9091/transmission/web/#open?url=<magnet_url>

That way I wouldn't need a separate server running which proxies the call under the hood.

My current plan looks like:
- deploy transmission on a pi
- deploy my custom application on same PI (probably use docker)
- open my custom application which let's me register custom protocol handler
- when I click on that link, my web app sends the details to the backend
- backend communicates with transmission and adds the magnet URL
- and redirects the frontend to transmission UI.

The best solution would be to of course transmission simply register the custom protocol handler, that way, when user clicks on magnet links, it automatically opens using that webpage and user just needs to confirm (just like the native experience)

But one more could work where going that url pre-fills the address (unless it already exists, I couldn't find any)

I'm happy to send a PR if interested.

**Comment by ckerr** [comment]
No, there's not currently any way to do that. All the RPC goes through POST right now.

Your use case makes sense; but still, it seems like a pretty uncommon use case. I think this might be better done as a third-party patch instead of going into `transmission/transmission`. If there's anything that can be done in libtransmission to help prevent 3rd party patches from breaking in the future, I'd be happy to review a PR for that.

**Comment by cyberhck** [comment]
I didn't mean add using a get, what I meant here is something like this: https://github.com/settings/tokens/new?description=ansible%20setup%20token&scopes=admin:gpg_key,admin:public_key see how it's prefilled? When we land to transmission UI, it could pre-fill the inputs, NOT actually add the torrent, that way user clicks on a magnet link, gets to transmission UI, then simply clicks on "Add" button, this is only web frontend change.

If you click on that link, it doesn't actually create the token, it still makes a post request, similarly for transmission, it'd just make a rpc call (only need to check if there's a url parameter on load, and if so, open the new dialog with that value pre-filled, no backend change should be necessary)

**Comment by ckerr** [comment]
Ah I see what you mean, OK. Thanks for link, that makes sense to me now.

I still think this would be a pretty uncommon use case so I'm undecided on this feature, but I'm no longer a definite "no". If you want to open a PR I'll review it. The two things I'm concerned about are related: size/complexity of this feature, and the cost of maintaining the code, e.g. if there are bug reports and you're not around and it's on me to fix it. Please, make the PR as simple and small as possible. :slightly_smiling_face:

**Comment by cyberhck** [comment]
Thanks, I'll have a look.

I'll run by you on my approach and let me know if you think if it makes sense.

On transmission.js's constructor I'll check if a ?add-torrent=<value> is present, if it's not, I simply move on, but if it is, I'll add the open dialog, and use browser apis to fill in the value.

I see transmission.js is getting bigger, I'll also give a think if there's a better way to do this.

Another thing I'll (try to) do but not sure if I'll be able to:
on this menu:
![image](https://user-images.githubusercontent.com/3640284/137963769-734b7382-cfc9-45ff-bdb4-4867b9542342.png)
add a button which basically adds the protocol handler (but it'll be added only if protocol handler is supported),

(this part is optional), but this is what let's anyone click on any magnet link and have it open transmission's UI. (Users can choose to use it or not, This is what I just recently built because I couldn't find it in transmission)

**Comment by ckerr** [comment]
You may want to check out what the web client looks like in `master` now before proceeding -- that context menu is completely gone now

**Comment by cyberhck** [comment]
ahh, will check that out this weekend. Now I understand why I can't find that :smile:

**Comment by cyberhck** [comment]
I was busy last weekend, I've put this on my todo-list, it might take some time though.

**Comment by killemov** [comment]
I have implemented a protocol handler in [Shift](https://github.com/killemov/Shift) and I came to the conclusion that the process should be as least intrusive as possible. You click on a magnet, Shift opens with "Uploading ...", the magnet uploads, the line is amended with "added" or "duplicate" and the browser returns to the previous page. The `start-added-torrents` setting is observed.
Discuss ...

**Comment by decyrus** [comment]
I had essentially the same use case: browse on a Mac, click a magnet link, and send it to Transmission running on a NAS or another server.

Rather than modifying the Transmission Web UI or deploying another service next to the daemon, I ended up building a small native macOS utility called [MagnetBridge](https://github.com/decyrus/magnet-bridge). It registers as the system magnet: handler, sends the link directly through Transmission’s RPC API, and can open the Web UI afterwards.

It doesn’t pre-fill Transmission’s add dialog as originally proposed here. Нhe torrent is submitted directly, either started immediately or added paused. It supports both Transmission 4.0.x and the newer 4.1+ RPC API.

`brew install --cask decyrus/tap/magnet-bridge`

Hopefully this is useful to other macOS users who land on this discussion with the same setup.
